xwiki-platform-core/xwiki-platform-oldcore/src/main/java/com/xpn/xwiki/internal/doc/XWikiAttachmentList.java

```diff
@@ -0,0 +1,250 @@
+     * @since 10.0RC1
+     */
+    @Override
+    public boolean add(XWikiAttachment attachment)
+    {
+        return set(attachment) != null;
+    }
+
+    /**
+     * Adds attachment to the list in order of filename.
+     * 
+     * @param index index is ignored as list is reordered based on filename
+     * @param attachment XWikiAttachment to add to the list
+     * @since 10.0RC1
+     */
+    @Override
+    public void add(int index, XWikiAttachment attachment)
+    {
+        add(attachment);
+    }
+
+    /**
+     * {@inheritDoc}
+     * 
+     * @since 10.0RC1
+     */
+    @Override
+    public void clear()
+    {
+        updateMap();
+    }
+
+    /**
+     * Adds all attachments to the list in order of filename.
+     * 
+     * @param index index is ignored as list is reordered based on filename
+     * @param c Collection that contains XWikiAttachment objects
+     * @since 10.0RC1
+     */
+    @Override
+    public boolean addAll(Collection<? extends XWikiAttachment> c)
+    {
+        for (XWikiAttachment x : c) {
+            map.put(x.getFilename(), x);
+            added(x);
+        }
+        updateMap();
+        return true;
+    }
+
+    /**
+     * Adds all attachments to the list in order of filename.
+     * 
+     * @param index index is ignored as list is reordered based on filename
+     * @param c Collection that contains XWikiAttachment objects
+     * @since 10.0RC1
+     */
+    @Override
+    public boolean addAll(int index, Collection<? extends XWikiAttachment> c)
+    {
+        return addAll(c);
+    }
+
+    /**
+     * {@inheritDoc}
+     * 
+     * @since 10.0RC1
+     */
+    @Override
+    public XWikiAttachment remove(int index)
+    {
+        XWikiAttachment removedAttachment = map.remove(this.list.get(index).getFilename());
+        if (removedAttachment != null) {
+            removedAttachment = super.remove(index);
+            document.setMetaDataDirty(true);
+        }
+        return removedAttachment;
+    }
+
+    /**
+     * Removes XWikiAttachment.
+     * 
+     * @param attachment XWikiAttachment to remove.
+     * @return true unless the attachment is not found
+     * @since 10.0RC1
+     */
+    @Override
+    public boolean remove(Object attachment)
+    {
+        XWikiAttachment xwikiAttachment = (XWikiAttachment) attachment;
+        if(map.put(xwikiAttachment.getFilename(), xwikiAttachment) == null)
+            return false;
+        map.remove(attachment);
+        updateMap();
+        return true;
+    }
+
+    /**
+     * Adds or replaces attachment with the same filename as the parameter.
+     * 
+     * @param attachment the attachment to add to the list
+     * @return the attachment that was added to the list in order of filename
+     * @since 10.0RC1
+     */
+    public XWikiAttachment set(XWikiAttachment attachment)
+    {
+        map.put(attachment.getFilename(), attachment);
+        added(attachment);
```

Comment:
You should probably do all that only if what map.put return is different from what you passed (i.e. something actually changed).